L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc

```diff
@@ -120,18 +152,98 @@ unsigned short HGCalTriggerTowerGeometryHelper::getTriggerTowerFromEtaPhi(const
   return l1t::HGCalTowerID(doNose_, zside, bin_eta, bin_phi).rawId();
 }
 
-unsigned short HGCalTriggerTowerGeometryHelper::getTriggerTower(const l1t::HGCalTriggerCell& thecell) const {
+std::unordered_map<unsigned short, float> HGCalTriggerTowerGeometryHelper::getTriggerTower(const l1t::HGCalTriggerCell& thecell) const {
+  std::unordered_map<unsigned short, float> towerIDandShares = {}; 
   unsigned int trigger_cell_id = thecell.detId();
   // NOTE: if the TC is not found in the map than it is mapped via eta-phi coords.
   // this can be considered dangerous (silent failure of the map) but it actually allows to save
   // memory mapping explicitly only what is actually needed
   auto tower_id_itr = cells_to_trigger_towers_.find(trigger_cell_id);
-  if (tower_id_itr != cells_to_trigger_towers_.end())
-    return tower_id_itr->second;
-
-  return getTriggerTowerFromEtaPhi(thecell.position().eta(), thecell.position().phi());
+  if (tower_id_itr != cells_to_trigger_towers_.end()){
+    towerIDandShares.insert({tower_id_itr->second, 1.0});
+    return towerIDandShares;
+  }
+  towerIDandShares.insert({getTriggerTowerFromEtaPhi(thecell.position().eta(), thecell.position().phi()), 1.0});
+  return towerIDandShares;
 }
 
-unsigned short HGCalTriggerTowerGeometryHelper::getTriggerTower(const l1t::HGCalTriggerSums& thesum) const {
-  return getTriggerTowerFromEtaPhi(thesum.position().eta(), thesum.position().phi());
+std::unordered_map<unsigned short, float> HGCalTriggerTowerGeometryHelper::getTriggerTower(const l1t::HGCalTriggerSums& thesum) const {
+  std::unordered_map<unsigned short, float> towerIDandShares = {}; 
+  if(!splitModuleSum_){
+    towerIDandShares.insert({getTriggerTowerFromEtaPhi(thesum.position().eta(), thesum.position().phi()), 1.0});
+    return towerIDandShares;
+  } else{
+    HGCalModuleDetId detid(thesum.detId());
+    int moduleU = detid.moduleU();
+    int moduleV = detid.moduleV();
+    int layer = detid.layer();
+    int sector = detid.sector();
+    int zside = detid.zside();
+
+    int subdet = -1;
+    int splitDivisor = -1;
+    if(detid.isHScintillator()){ //FIXME HFNose not handled
+      subdet = 2;
+      layer += 28;
+      splitDivisor = splitDivisorScint_; 
+    } else if(detid.isEE()){
+      subdet = 0;
+      splitDivisor = splitDivisorSilic_;
+    } else if(detid.isHSilicon()) {
+      subdet = 1;
+      layer += 28;
+      splitDivisor = splitDivisorSilic_;
+    }
+    
+    auto module_id_itr = modules_to_trigger_towers_.find(std::to_string(subdet) + " " + 
+                                                         std::to_string(layer) + " " + 
+                                                         std::to_string(moduleU) + " " + 
+                                                         std::to_string(moduleV)
+                                                         );
```

Comment:
Should not have `string` manipulation here. This is costly. Should rather use an integer id that packs the subdet, layer, moduleU, moduleV information.
You can have a look at what is done e.g. here:
https://github.com/PFCal-dev/cmssw/blob/f82947aa34c45df7dd9fa3e96884208390464d63/L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc#L877-L885
https://github.com/PFCal-dev/cmssw/blob/f82947aa34c45df7dd9fa3e96884208390464d63/L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc#L526-L527